### PR TITLE
857 certificates handling over an unsecure connection in ocpp201 should not be allowed

### PR DESCRIPTION
**Pull request by shingoxx222**
## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] I have made corresponding changes to the documentation
- [x] If OCPP 2.0.1: I have updated the [OCPP 2.0.1 status document](https://github.com/EVerest/libocpp/tree/main/doc/ocpp_201_status.md)
- [x] I read the [contribution documentation](https://github.com/EVerest/EVerest/blob/main/CONTRIBUTING.md) and made sure that my changes meet its requirements

### CODE REVIEW

**Comment by shingoxx222** [comment]
test result:
image

**Review by AssemblyJohn — COMMENTED** [review]
At an initial look, it looks good to me.

File: include/ocpp/v201/smart_charging.hpp [hunk 1]
```diff
@@ -26,6 +26,7 @@ enum class ProfileValidationResultEnum {
     Valid,
     EvseDoesNotExist,
     ExistingChargingStationExternalConstraints,
+    InvalidSecurityProfile,
```
Comment (marcemmers) [inline include/ocpp/v201/smart_charging.hpp:29]: Since this is unrelated to smart charging we should not add this to a result enum of smart charging

File: lib/ocpp/v201/charge_point.cpp [hunk 2]
```diff
@@ -3594,18 +3594,26 @@ void ChargePoint::handle_install_certificate_req(Call<InstallCertificateRequest>
...
-        msg.certificate.get(), ocpp::evse_security_conversions::from_ocpp_v201(msg.certificateType));
-    response.status = ocpp::evse_security_conversions::to_ocpp_v201(result);
-
-    if (response.status == InstallCertificateStatusEnum::Accepted) {
-        const auto& security_event = ocpp::security_events::RECONFIGURATIONOFSECURITYPARAMETERS;
-        std::string tech_info =
-            "Installed certificate: " + conversions::install_certificate_use_enum_to_string(msg.certificateType);
-        this->security_event_notification_req(CiString<50>(security_event), CiString<255>(tech_info), true,
-                                              utils::is_critical(security_event));
+    if ((msg.certificateType == InstallCertificateUseEnum::CSMSRootCertificate ||
```
Comment (marcemmers) [inline lib/ocpp/v201/charge_point.cpp:3597]: Allowing installation of `CSMSRootCertificate` and `ManufacturerRootCertificate` should be configurable, not hardcoded. I think preferably they should be configurable separately from each other.

Default should also allow this since that is what the current implementation does.

**Review by marcemmers — CHANGES_REQUESTED** [review]

File: include/ocpp/v201/smart_charging.hpp (hunk 1 above)
Comment (shingoxx222) [reply include/ocpp/v201/smart_charging.hpp]: using hardcoded string now

File: lib/ocpp/v201/charge_point.cpp (hunk 2 above)
Comment (shingoxx222) [reply lib/ocpp/v201/charge_point.cpp]: done, I make them `ReadWrite` and `not required`

File: config/v201/component_config/standardized/SecurityCtrlr.json [hunk 3]
```diff
@@ -177,6 +177,38 @@
       "maximum": 3,
       "default": "1",
       "type": "integer"
+    },
+    "AllowCSMSRootCertificateInstallWhenLowSecurityProfile": {
+      "variable_name": "AllowCSMSRootCertificateInstallWhenLowSecurityProfile",
```
Comment (marcemmers) [inline config/v201/component_config/standardized/SecurityCtrlr.json:182]: "Low" security profile leaves room for interpretation, it could be a bit more explicit still ;)

File: lib/ocpp/v201/charge_point.cpp [hunk 4]
```diff
@@ -3588,24 +3588,52 @@ void ChargePoint::handle_get_installed_certificate_ids_req(Call<GetInstalledCert
     this->send<GetInstalledCertificateIdsResponse>(call_result);
 }
 
+bool ChargePoint::should_reject_certificate_install(InstallCertificateUseEnum cert_type) const {
```
Comment (marcemmers) [inline lib/ocpp/v201/charge_point.cpp:3591]: Nice to make a function of this to clean up the handler :+1: 

I do always have a difficult time with the double negatives especially with `value_or` in the mix. Would it be an option to change this to `should_accept_certificate_install` instead?

File: config/v201/component_config/standardized/SecurityCtrlr.json [hunk 5]
```diff
@@ -177,6 +177,38 @@
...
+      "characteristics": {
+        "supportsMonitoring": true,
+        "dataType": "boolean"
+      },
+      "attributes": [
+        {
+          "type": "Actual",
+          "mutability": "ReadWrite"
```
Comment (marcemmers) [inline config/v201/component_config/standardized/SecurityCtrlr.json:190]: What is the reason you've chosen ReadWrite here? Keep in mind the reason for this change is to prevent parties to install wrong certificates.

**Review by marcemmers — CHANGES_REQUESTED** [review]

File: lib/ocpp/v201/charge_point.cpp [hunk 6]
```diff
@@ -3588,24 +3588,52 @@ void ChargePoint::handle_get_installed_certificate_ids_req(Call<GetInstalledCert
...
-        const auto& security_event = ocpp::security_events::RECONFIGURATIONOFSECURITYPARAMETERS;
-        std::string tech_info =
-            "Installed certificate: " + conversions::install_certificate_use_enum_to_string(msg.certificateType);
-        this->security_event_notification_req(CiString<50>(security_event), CiString<255>(tech_info), true,
-                                              utils::is_critical(security_event));
+    if (should_reject_certificate_install(msg.certificateType)) {
+        response.status = InstallCertificateStatusEnum::Rejected;
+        response.statusInfo = StatusInfo();
+        response.statusInfo->reasonCode = "LowSecurityProfile";
+        response.statusInfo->additionalInfo = "SecurityProfileTooLowForCertificateInstall";
```
Comment (marcemmers) [inline lib/ocpp/v201/charge_point.cpp:3624]: This one could also be improved a bit. For example "Installation of certificate is not allowed when security profile is less then 2"

File: lib/ocpp/v201/ctrlr_component_variables.cpp [hunk 7]
```diff
@@ -1116,6 +1116,20 @@ const RequiredComponentVariable& SecurityProfile = {
         "SecurityProfile",
     }),
 };
+const ComponentVariable& AllowCSMSRootCertificateInstallWhenLowSecurityProfile = {
+    ControllerComponents::SecurityCtrlr,
```
Comment (marcemmers) [inline lib/ocpp/v201/ctrlr_component_variables.cpp:1120]: They should be part of the InternalCtrlr since they are configuration items used internally

Comment (shingoxx222) [reply lib/ocpp/v201/ctrlr_component_variables.cpp]: done

File: lib/ocpp/v201/charge_point.cpp (hunk 6 above)
Comment (shingoxx222) [reply lib/ocpp/v201/charge_point.cpp]: made it `CertificateInstallationNotAllowedWhenSecurityProfile1`

File: config/v201/component_config/standardized/SecurityCtrlr.json (hunk 5 above)
Comment (shingoxx222) [reply config/v201/component_config/standardized/SecurityCtrlr.json]: I thought `ReadOnly` is the same as unmutable. now it is `ReadOnly`

File: lib/ocpp/v201/charge_point.cpp (hunk 4 above)
Comment (shingoxx222) [reply lib/ocpp/v201/charge_point.cpp]: done

File: config/v201/component_config/standardized/SecurityCtrlr.json (hunk 3 above)
Comment (shingoxx222) [reply config/v201/component_config/standardized/SecurityCtrlr.json]: it is now `AllowCSMSRootCertInstallWhenSecurityProfile1`

**Review by marcemmers — APPROVED** [review]
